index/segment/fst/reader.go

```diff
@@ -0,0 +1,471 @@
+// we're forced to copy these massive slices every time. Tracking this under
+// https://github.com/m3db/m3ninx/issues/66
+func (r *reader) Fields() ([][]byte, error) {
+	r.RLock()
+	defer r.RUnlock()
+	if r.closed {
+		return nil, errReaderClosed
+	}
+
+	return r.allKeys(r.fieldsFST)
+}
+
+func (r *reader) Terms(field []byte) ([][]byte, error) {
+	r.RLock()
+	defer r.RUnlock()
+	if r.closed {
+		return nil, errReaderClosed
+	}
+
+	termsFST, err := r.retrieveTermsFST(field)
+	if err != nil {
+		return nil, err
+	}
+
+	fstCloser := x.NewSafeCloser(termsFST)
+	defer fstCloser.Close()
+
+	terms, err := r.allKeys(termsFST)
+	if err != nil {
+		return nil, err
+	}
+
+	if err := fstCloser.Close(); err != nil {
+		return nil, err
+	}
+
+	return terms, nil
+}
+
+func (r *reader) MatchTerm(field []byte, term []byte) (postings.List, error) {
+	r.RLock()
+	defer r.RUnlock()
+	if r.closed {
+		return nil, errReaderClosed
+	}
+
+	termsFST, err := r.retrieveTermsFST(field)
+	if err != nil {
+		return nil, err
+	}
+	fstCloser := x.NewSafeCloser(termsFST)
+	defer fstCloser.Close()
+
+	postingsOffset, exists, err := termsFST.Get(term)
+	if err != nil {
+		return nil, err
+	}
+
+	if !exists {
+		// i.e. we don't know anything about the term, so can early return an empty postings list
+		return r.opts.PostingsListPool.Get(), nil
+	}
+
+	pl, err := r.retrievePostingsList(postingsOffset)
+	if err != nil {
+		return nil, err
+	}
+
+	if err := fstCloser.Close(); err != nil {
+		return nil, err
+	}
+
+	return pl, nil
+}
+
+func (r *reader) MatchRegexp(field []byte, regexp []byte, compiled *regexp.Regexp) (postings.List, error) {
+	re, err := vregex.New(string(regexp))
+	if err != nil {
+		return nil, err
+	}
+
+	termsFST, err := r.retrieveTermsFST(field)
+	if err != nil {
+		return nil, err
+	}
+
+	var (
+		fstCloser     = x.NewSafeCloser(termsFST)
+		pl            = r.opts.PostingsListPool.Get()
+		iter, iterErr = termsFST.Search(re, minByteKey, maxByteKey)
+		iterCloser    = x.NewSafeCloser(iter)
+	)
+	defer func() {
+		iterCloser.Close()
+		fstCloser.Close()
+	}()
+
+	for {
+		if iterErr == vellum.ErrIteratorDone {
+			break
+		}
+
+		if iterErr != nil {
+			return nil, iterErr
+		}
+
+		_, postingsOffset := iter.Current()
+		nextPl, err := r.retrievePostingsList(postingsOffset)
+		if err != nil {
+			return nil, err
+		}
+		if err := pl.Union(nextPl); err != nil {
+			return nil, err
+		}
+
+		iterErr = iter.Next()
+	}
+
+	if err := iterCloser.Close(); err != nil {
+		return nil, err
+	}
+
+	if err := fstCloser.Close(); err != nil {
+		return nil, err
+	}
+
+	return pl, nil
+}
+
+func (r *reader) MatchAll() (postings.MutableList, error) {
+	return nil, errNotImplemented
+}
+
+func (r *reader) Docs(pl postings.List) (doc.Iterator, error) {
+	return nil, errNotImplemented
+}
+
+func (r *reader) AllDocs() (doc.Iterator, error) {
+	return nil, errNotImplemented
+}
+
+func (r *reader) retrievePostingsList(postingsOffset uint64) (postings.List, error) {
```

Comment:
nit: is it worth signaling that this must be called the `RLock` in the name? Same with those below?

Reply:
done